monai/networks/nets/basic_unet.py

```diff
@@ -0,0 +1,164 @@
+# Copyright 2020 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import torch.nn as nn
+
+from monai.networks.blocks import Convolution, UpSample
+from monai.networks.layers.factories import Conv, Pool
+from monai.utils import ensure_tuple_rep
+
+__all__ = ["BasicUNet", "BasicUnet", "Basicunet"]
+
+
+class TwoConv(nn.Sequential):
+    """two convolutions."""
+
+    def __init__(self, dim, in_chns, out_chns, act, norm, dropout):
+        super().__init__()
+
+        conv_0 = Convolution(dim, in_chns, out_chns, act=act, norm=norm, dropout=dropout, padding=1)
+        conv_1 = Convolution(dim, out_chns, out_chns, act=act, norm=norm, dropout=dropout, padding=1)
+        self.add_module("conv_0", conv_0)
+        self.add_module("conv_1", conv_1)
+
+
+class Down(nn.Sequential):
+    """maxpooling downsampling and two convolutions."""
+
+    def __init__(self, dim, in_chns, out_chns, act, norm, dropout):
+        super().__init__()
+
+        max_pooling = Pool["MAX", dim](kernel_size=2)
+        convs = TwoConv(dim, in_chns, out_chns, act, norm, dropout)
+        self.add_module("max_pooling", max_pooling)
+        self.add_module("convs", convs)
+
+
+class UpCat(nn.Module):
+    """upsampling, concatenation with the encoder feature map, two convolutions"""
+
+    def __init__(self, dim, in_chns, cat_chns, out_chns, act, norm, dropout, upsample="deconv", halves=True):
+        super().__init__()
+
+        up_chns = in_chns // 2 if halves else in_chns
+        self.upsample = UpSample(dim, in_chns, up_chns, 2, mode=upsample)
+        self.convs = TwoConv(dim, cat_chns + up_chns, out_chns, act, norm, dropout)
+
+    def forward(self, x, x_e):
+        """x: features to be upsampled; x_e: features from the encoder"""
+        x_0 = self.upsample(x)
+
+        # handling spatial shapes due to the 2x maxpooling with odd edge lengths.
+        dimensions = x.ndim - 2
+        sp = [0] * (dimensions * 2)
+        for i in range(dimensions):
+            if x_e.shape[-i - 1] != x_0.shape[-i - 1]:
+                sp[i * 2 + 1] = 1
+        if sum(sp) != 0:
+            x_0 = torch.nn.functional.pad(x_0, sp, "replicate")
+
+        x = self.convs(torch.cat([x_e, x_0], dim=1))  # input channels: (cat_chns + up_chns)
+        return x
+
+
+class BasicUNet(nn.Module):
+    def __init__(
+        self,
+        dimensions=3,
```

Comment:
Here the input parameters may need to add types, just as other networks we implemented. Like:
```
dimensions: int = 3,
in_channels: int = 1,
out_channels: int = 2,
...
```

Reply:
Same for other classes: `TwoConv`, `Down`, `UpCat`